Tools/sfparse.1
.\"Copyright (C) 2005 Free Software Foundation, Inc.
.TH SFPARSE "1" "February 2004" "GNUstep" "GNUstep System Manual"
.SH NAME
sfparse \- string file parse

.SH SYNOPSIS
.B sfparse
[ \fI--utf8\fR | \fI--unicode\fR ] \fIfilename\fR

.SH DESCRIPTION
.P
This tool has two different functions.  When invoked with just a filename as
argument, it checks that a file is a valid strings-file, which basically
amounts to seeing whether the file is in proper serialized NSDictionary format
(type "\fIdefaults plist\fR" for info on this).
.P
On the other hand, if one of the two options is given, no such checking is
performed, but the file IS converted into the named encoding.

.SH OPTIONS
.P
\fB\-\-unicode\fR    - convert an ASCII or UTF8 file to Unicode
.P
\fB\-\-utf8\fR       - convert an ASCII or Unicode to UTF8

.SH HISTORY
.RS 0
Written in 1999.
.P
This manual page first appeared in gnustep-base 1.9.2 (March 2004).
.P
.SH AUTHORS
.B sfparse
was written by Richard Frith-McDonald <rfm@gnu.org>


NEWS
News
****

The currently released version of the library is `1.10.3'.

See the <ReleaseNotes.html> document for more information.

Noteworthy changes in version `1.10.3'
======================================

This version includes a few minor bug fixes.

Noteworthy changes in version `1.10.2'
======================================

This version mostly includes minor fixes and updates.

   * The definition of NSContainsRect has changed to correspond with the
     current Mac OS X/Cocoa def - the sides of the bRect can touch
     aRect.

   * Add support for systems that support unichar file paths (e.g.
     Windows).

   * Use a proper one-to-one abbreviation dictionary for NSTimeZone.

Noteworthy changes in version `1.10.1'
======================================

This version mostly includes minor fixes and updates.

   * Runtime discovery of the library version is possible.

   * NSXMLParser class added. Not completely implemented.

   * NSArrays makeObjectsPerformSelector iterates in same order as
     MacOS X.

   * Designated initializers for NSArray, NSDictionary, NSSet, and
     NSString have been changed for MacOS X compatibility.

Noteworthy changes in version `1.10.0'
======================================

Note the interface version of the library has changed so that apps,
tools and libraries the use the base library need to be recompiled to
use it.

   * A large majority of the library classes have been documented
     thanks to the efforts of Andrew Ruder

   * URL classes support persistant connections

   * Mac OSX XML compatibility fixes.

Noteworthy changes in version `1.9.2'
=====================================

   * GSMime parsing ignores extraneous data

   * New log functions GSOnceFlag and GSOnceMLog

   * New class NSError

   * Multiple new function in GSObjCRuntime

   * NSProtocolChecker rewritten

   * autogsdoc support added for building frames

   * Binary incompatibility:  NSUnarchiver, GSIMapTable have new ivars
     added

Noteworthy changes in version `1.9.1'
=====================================

   * Default string encoding taken from system nl_langinfo if not set
     with GNUSTEP_STRING_ENCODING.

   * NSKeyedArchiver/NSKeyedUnarchiver classes added.

   * NSPropertyLists class added, also decodes Mac OS X binary propery
     lists.

Noteworthy changes in version `1.9.0'
=====================================

   * Lazy locking implemented (see GSLock documentation)

   * MacOSX PB files updated to 10.3

   * DO retain/release implementation simplified and also handles some
     complex DO retain/release problems.

   * GNUSTEP_STRING_ENCODING now supports any valid iconv name.

   * gdomap uid/gid fixes

   * NSString subclass heirarchy reorganized to fix problems with
     copying and ownership of data.

Noteworthy changes in version `1.8.0'
=====================================

Read the NEWS file for a complete list of changes since the last stable
version. The major change is that the header files (both in the package
and when installed) are in different locations. Also, due to filesystem
changes in gnustep-make, several components of gnustep-base are located
in different locations. Generally this should not affect the
compilation or running of applications and tools.

Noteworthy changes in version `1.7.4'
=====================================

   * Added Korean encoding

   * Updated use of ObjCRuntime functions.

Noteworthy changes in version `1.7.3'
=====================================

Note in this version there have been large changes to the location of
header files (both in the package and when installed). It's possible
(though unlikely) that certain applications that rely on GNUstep
specific headers may not compile because of this.

   * A number of Protocol and DO fixes.

   * New TraditionalChinese language.

Noteworthy changes in version `1.7.2'
=====================================

   * NSUndoManager improvements

   * New NSMessagePort implementation.

   * Private GSTcpPort becomre NSSocketPort.

   * Improvements to handle selectors better over remote connections.

Noteworthy changes in version `1.7.1'
=====================================

Bug fixes.

Noteworthy changes in version `1.7.0'
=====================================

See gnustep-make for comments on the filesystem change. Some components
of gnustep-base are now stored in different directories and in some
cases you should use different methods to look for them. Use
+pathForLibraryResource:ofType:inDirectory: instead of
+pathForGNUstepResource:. Also try using +bundleForLibrary:.

   gnustep-base is now documentated as OpenStep complient. Other
changes:

   * Add GSApplicationSupportDirectory for
     NSSearchPathForDirectoriesInDomains.

   * Added man page for gdomap

Noteworthy changes in version `1.6.0'
=====================================

   * More MinGW support

   * New MacOSX compatible methods.

   * The Base Additions library compiles on native OSX.

   * autogsdoc produces better, cleaner docs.

   * Built-in garbage collecting classes.

   * Any many many bug fixes and minor improvements.

Noteworthy changes in version `1.5.1'
=====================================

   * Port NSThread to MingW

   * Better framework support in NSBundle

   * Added Thai string encodings.

   * New makeImmutableCopyOnFail: method.

   * Lots of autogsdoc improvements, Updated gsdoc DTD to 1.0 version.

   * Added GSPrintf for logging to stdout.

   * Warning logs selectable at runtime.

   * NSTimeZone code rewritten for speed.

Noteworthy changes in version `1.5.0'
=====================================

   * New MacOSX methods implemented (NSString, NSArray, NSObject)

   * Fixed bug that occationally deleted user defaults.

   * Tcp connections use runloop in NSConnectionReplyMode.

Noteworthy changes in version `1.4.0'
=====================================

   * gdomap - security bug fixes.

Noteworthy changes in version `1.3.4'
=====================================


Tools/sfparse.m
/** This tool checks that a file is a valid strings-file
   Copyright (C) 1999 Free Software Foundation, Inc.

   Written by:  Richard Frith-Macdonald <richard@brainstorm.co.uk>
   Created: February 1999

   This file is part of the GNUstep Project

   This library is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.

   You should have received a copy of the GNU General Public
   License along with this library; see the file COPYING.LIB.
   If not, write to the Free Software Foundation,
   51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.

   */

#include "config.h"
#include	<Foundation/NSArray.h>
#include	<Foundation/NSData.h>
#include	<Foundation/NSException.h>
#include	<Foundation/NSString.h>
#include	<Foundation/NSProcessInfo.h>
#include	<Foundation/NSUserDefaults.h>
#include	<Foundation/NSDebug.h>
#include	<Foundation/NSAutoreleasePool.h>
#ifdef NeXT_Foundation_LIBRARY
#include "GNUstepBase/GSCategories.h"
#endif

int
convert_unicode(NSArray *args)
{
  unsigned int i;

  for (i = 2; i < [args count]; i++)
    {
      NSString	*file = [args objectAtIndex: i];

      NS_DURING
	{
	  NSData        *data;
	  NSString	*myString;
	  NSString      *output;

	  data = [NSData dataWithContentsOfFile: file];
	  myString = [[NSString alloc] initWithData: data
					   encoding: NSUTF8StringEncoding];
	  AUTORELEASE(myString);
	  if ([myString length] == 0)
	    {
	      myString = [[NSString alloc] initWithData: data
		encoding: [NSString defaultCStringEncoding]];
	    }
	  output = [[file lastPathComponent]
	    stringByAppendingPathExtension: @"unicode"];
	  data = [myString dataUsingEncoding: NSUnicodeStringEncoding];
	  [data writeToFile: output atomically: YES];
	}
      NS_HANDLER
	{
	  GSPrintf(stderr, @"Converting '%@' - %@\n", file,
	    [localException reason]);
	  return 1;
	}
      NS_ENDHANDLER
    }
  return 0;
}

int
convert_utf8(NSArray *args)
{
  unsigned int i;

  for (i = 2; i < [args count]; i++)
    {
      NSString	*file = [args objectAtIndex: i];

      NS_DURING
	{
	  NSData        *data;
	  NSString	*myString;
	  NSString      *output;

	  myString = [NSString stringWithContentsOfFile: file];
	  output = [[file lastPathComponent]
	    stringByAppendingPathExtension: @"utf8"];
	  data = [myString dataUsingEncoding: NSUTF8StringEncoding];
	  [data writeToFile: output atomically: YES];
	}
      NS_HANDLER
	{
	  GSPrintf(stderr, @"Converting '%@' - %@\n", file,
	    [localException reason]);
	  return 1;
	}
      NS_ENDHANDLER
    }
  return 0;
}


/** <p>
 This tool checks that a file is a valid strings-file, and can also convert
 files to Unicode or UTF-8.  If given the '<code>--unicode</code>' option
 it converts an ASCII or UTF-8 file to unicode.  If given the
 '<code>--utf8</code>' option is converts an ASCII or unicode file to UTF-8.
 </p> */
int
main(int argc, char** argv, char **env)
{
  NSAutoreleasePool	*pool;
  NSProcessInfo		*proc;
  NSArray		*args;
  unsigned		i;
  int			retval = 0;

#ifdef GS_PASS_ARGUMENTS
  [NSProcessInfo initializeWithArguments:argv count:argc environment:env];
#endif
  pool = [NSAutoreleasePool new];
  proc = [NSProcessInfo processInfo];
  if (proc == nil)
    {
      GSPrintf(stderr, @"defaults: unable to get process information!\n");
      [pool release];
      exit(EXIT_FAILURE);
    }

  args = [proc arguments];

  if ([args count] <= 1 || [[args objectAtIndex: 1] isEqual: @"--help"]
      || [[args objectAtIndex: 1] isEqual: @"-h"])
    {
      printf("Usage: sfparse [--utf8] filename.\n");
      printf("--unicode    - convert an ASCII or UTF8 file to Unicode\n");
      printf("--utf8       - convert an ASCII or Unicode to UTF8\n");
    }
  else if ([[args objectAtIndex: 1] isEqual: @"--unicode"])
    {
      retval = convert_unicode(args);
    }
  else if ([[args objectAtIndex: 1] isEqual: @"--utf8"])
    {
      retval = convert_utf8(args);
    }
  else
    {
      for (i = 1; i < [args count]; i++)
	{
	  NSString	*file = [args objectAtIndex: i];

	  NS_DURING
	    {
	      NSString	*myString;
	      id		result;

	      myString = [NSString stringWithContentsOfFile: file];
	      result = [myString propertyListFromStringsFileFormat];
	      if (result == nil)
		GSPrintf(stderr, @"Parsing '%@' - nil property list\n", file);
	      else if ([result isKindOfClass: [NSDictionary class]] == YES)
		GSPrintf(stderr, @"Parsing '%@' - seems ok\n", file);
	      else
		GSPrintf(stderr, @"Parsing '%@' - unexpected class - %@\n",
		  file, [[result class] description]);
	    }
	  NS_HANDLER
	    {
	      GSPrintf(stderr, @"Parsing '%@' - %@\n", file,
		[localException reason]);
	      retval = 1;
	    }
	  NS_ENDHANDLER
	}
    }
  [pool release];
  return retval;
}
